public/app/features/plugins/pluginPreloader.test.ts

```diff
@@ -0,0 +1,264 @@
+        showErrorAlert: true,
+      });
+      expect(getPluginSettingsMock).toHaveBeenNthCalledWith(2, 'plugin-2', {
+        showErrorAlert: true,
+      });
+      expect(importAppPluginMock).toHaveBeenCalledTimes(2);
+      expect(importAppPluginMock).toHaveBeenNthCalledWith(1, mockPluginMeta1);
+      expect(importAppPluginMock).toHaveBeenNthCalledWith(2, mockPluginMeta2);
+    });
+
+    it('should not preload already preloaded plugins', async () => {
+      const appConfigs = [
+        createMockAppPluginConfig({
+          id: 'plugin-1',
+          path: '/path/to/plugin1',
+          version: '1.0.0',
+        }),
+      ];
+
+      const mockPluginMeta = createMockPluginMeta({
+        id: 'plugin-1',
+        name: 'Plugin 1',
+        type: PluginType.app,
+      });
+
+      getPluginSettingsMock.mockResolvedValue(mockPluginMeta);
+      importAppPluginMock.mockResolvedValue({} as any);
+
+      await preloadPlugins(appConfigs);
+      await preloadPlugins(appConfigs);
+      await preloadPlugins(appConfigs);
+
+      expect(getPluginSettingsMock).toHaveBeenCalledTimes(1);
+      expect(importAppPluginMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('should not preload plugins twice, even if the initial has not finished yet', async () => {
+      const appConfigs = [
+        createMockAppPluginConfig({
+          id: 'plugin-1',
+          path: '/path/to/plugin1',
+          version: '1.0.0',
+        }),
+      ];
+
+      const mockPluginMeta = createMockPluginMeta({
+        id: 'plugin-1',
+        name: 'Plugin 1',
+        type: PluginType.app,
+      });
+
+      getPluginSettingsMock.mockImplementation(
+        () => new Promise((resolve) => setTimeout(() => resolve(mockPluginMeta), 100))
+      );
+      importAppPluginMock.mockImplementation(() => new Promise((resolve) => setTimeout(() => resolve({} as any), 100)));
+
+      // We are not awaiting the first call, so the second one kicks off before the first one has finished
+      preloadPlugins(appConfigs);
+      await preloadPlugins(appConfigs);
+
+      expect(getPluginSettingsMock).toHaveBeenCalledTimes(1);
+      expect(importAppPluginMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('should not preload plugins twice, even if the upcoming calls have a different set of plugins', async () => {
+      const appConfig1 = createMockAppPluginConfig({
+        id: 'plugin-1',
+        path: '/path/to/plugin1',
+        version: '1.0.0',
+      });
+
+      const appConfig2 = createMockAppPluginConfig({
+        id: 'plugin-2',
+        path: '/path/to/plugin2',
+        version: '2.0.0',
+      });
+
+      const mockPluginMeta1 = createMockPluginMeta({
+        id: 'plugin-1',
+        name: 'Plugin 1',
+        type: PluginType.app,
+      });
+
+      const mockPluginMeta2 = createMockPluginMeta({
+        id: 'plugin-2',
+        name: 'Plugin 2',
+        type: PluginType.app,
+      });
+
+      getPluginSettingsMock
```

Comment:
Maybe we can utilize jest's Timer mocks like jest.useFakeTimers() instead of setTimeout?